syz-verifier/verifier.go

```diff
@@ -31,24 +36,180 @@ type Verifier struct {
+	reasons           map[*prog.Syscall]string
+	reportReasons     bool
+	stats             *Stats
+	statsWrite        io.Writer
+	newEnv            bool
+	reruns            int
+
+	// We use single queue for every kernel environment.
+	tasksMutex     sync.Mutex
+	onTaskAdded    *sync.Cond
+	kernelEnvTasks [][]*ExecTaskQueue
+}
+
+func (vrf *Verifier) Init() {
+	vrf.progGeneratorInit.Add(1)
+
+	vrf.onTaskAdded = sync.NewCond(&vrf.tasksMutex)
+
+	vrf.kernelEnvTasks = make([][]*ExecTaskQueue, len(vrf.pools))
+	for i := range vrf.kernelEnvTasks {
+		vrf.kernelEnvTasks[i] = make([]*ExecTaskQueue, EnvironmentsCount)
+	}
+
+	for i := range vrf.kernelEnvTasks {
+		for j := range vrf.kernelEnvTasks[i] {
+			vrf.kernelEnvTasks[i][j] = MakeExecTaskQueue()
+		}
+	}
+
+	srv, err := startRPCServer(vrf)
+	if err != nil {
+		log.Fatalf("failed to initialise RPC server: %v", err)
+	}
+	vrf.srv = srv
+}
+
+func (vrf *Verifier) StartProgramsAnalysis() {
+	go func() {
+		vrf.progGeneratorInit.Wait()
+
+		type AnalysisResult struct {
+			Diff []*ExecResult
+			Prog *prog.Prog
+		}
+
+		results := make(chan *AnalysisResult)
+		go func() {
+			for result := range results {
+				if result.Diff != nil {
+					vrf.SaveDiffResults(result.Diff, result.Prog)
+				}
+			}
+		}()
+
+		sem := semaphore.NewWeighted(int64(100))
+		ctx := context.Background()
+		for {
+			sem.Acquire(ctx, 1)
+			go func(prog *prog.Prog) {
+				defer sem.Release(1)
+				results <- &AnalysisResult{
+					vrf.TestProgram(prog),
+					prog,
+				}
+			}(vrf.generate())
+		}
+	}()
+}
+
+func (vrf *Verifier) GetRunnerTask(kernel int, existing EnvDescr) *rpctype.ExecTask {
+	vrf.tasksMutex.Lock()
+	defer vrf.tasksMutex.Unlock()
+
+	for {
+		for env := existing; env >= AnyEnvironment; env-- {
+			if task, ok := vrf.kernelEnvTasks[kernel][env].PopTask(); ok {
+				return task.ToRPC()
+			}
+		}
+
+		vrf.onTaskAdded.Wait()
+	}
+}
+
+func PutExecResult(result *ExecResult) {
+	go func() {
+		c := GetExecResultChan(result.ExecTaskID)
+		c <- result
+	}()
+}
+
+// TestProgram return the results slice if some exec diff was found.
+func (vrf *Verifier) TestProgram(prog *prog.Prog) (result []*ExecResult) {
+	steps := []EnvDescr{
+		NewEnvironment,
+		NewEnvironment,
+	}
+
+	defer atomic.AddInt64(&vrf.stats.TotalProgs, 1)
+
+	for i, env := range steps {
+		stepRes, err := vrf.Run(prog, env)
+		if err != nil {
+			return
+		}
+		vrf.AddCallsExecutionStat(stepRes, prog)
+		if stepRes[0].IsEqual(stepRes[1]) {
+			if i != 0 {
+				atomic.AddInt64(&vrf.stats.FlakyProgs, 1)
+			}
+			return
+		}
+		if i == len(steps)-1 {
+			atomic.AddInt64(&vrf.stats.MismatchingProgs, 1)
+			return stepRes
+		}
+	}
+	return
+}
+
+// Run sends the program for verification to execution queues and return
+// result once it's ready.
+// In case of time-out, return (nil, error).
+func (vrf *Verifier) Run(prog *prog.Prog, env EnvDescr) (result []*ExecResult, err error) {
+	result = make([]*ExecResult, 2)
```

Comment:
s/2/len(vrf.kernelEnvTasks)/

Reply:
Currently use only 2 kernels, but remember that it could be extended to support N.
As a result, some code supports 2 and some code supports N.
Agree, at the functions level this assumption should be consistent - 2 or N.

Reply:
Currently use only 2 kernels, but remember that it could be extended to support N.
As a result, some code supports 2 and some code supports N.
Agree, at the functions level this assumption should be consistent - 2 or N.

Reply:
If we use 2 somewhere and it's not easy to change to len(vrf.kernelEnvTasks), then we should use some named const to make it easier to fix up in future.